### PR TITLE
[RLlib] Upgrade gym to 0.23.0

### LINKED ISSUE #23286: [Bug] [RLlib] module 'gym.wrappers' has no attribute 'Monitor'
## Ray Component
RLlib

## Problem description
gym version 0.23.1 has removed `Monitor` from wrappers, so `ray/rllib/env/utils.py` do not work!

## Issue Severity
Medium: I solved the problem installing gym version 0.21.0

## Code reference
https://github.com/ray-project/ray/blob/f400b4333a3f9008ebfc069d1e6c6812f996e995/rllib/env/utils.py#L59
ray/rllib/evaluate.py also uses gym_wrappers.Monitor

### PR DESCRIPTION
**Pull request by avnishn**
## Why are these changes needed?
A PR that removes the gym monitor allowing us to upgrade to gym 0.23.0

Currently, this PR doesn't support the newest gym feature which is reset options. 

The reset options to be supported are:
- `seed`: `int`: which seeds the gym environment's internal RNG upon reset. This only works if the environment uses the internal RNG object.
- `return_info`: `bool`: which returns a boolean representing whether an `env_infos` dictionary should be returned with the reset observation
- `options`: `dict` which is a dictionary that can be passed so that the environment can perform some reset behavior e.g. curriculum learning, meta/multi-task RL. 

I'm planning to add these options to our `try_reset` function.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by duburcqa** [comment]
I merged the necessary modifications in `gym` for `dm-tree` to be able to operate natively on `gym.spaces`, including dict and tuples. It was added to `gym>=0.23.0`, so [get_base_struct_from_space](https://github.com/ray-project/ray/blob/990b0ec93420973939c0d68284a774cfc6d4798f/rllib/utils/spaces/space_utils.py#L40) can be removed everywhere ! It would improve performance and clean up the code :)

**Review by ArturNiederfahrenhorst — COMMENTED** [review]
Could not find anything particular but must add that this is the first time I read the monitor/recording code and I'm yet to learn about it's usage in RLlib.

File: rllib/agents/trainer.py [hunk 1]
```diff
@@ -237,13 +237,6 @@
     # a) handles window generation and rendering itself (returning True) or
     # b) returns a numpy uint8 image of shape [height x width x 3 (RGB)].
     "render_env": False,
-    # If True, stores videos in this relative directory inside the default
```
[the remark this answers is not in the document]
Comment (zseymour) [reply rllib/agents/trainer.py:241]: Couldn't the option be kept and `Monitor` replaced with the new `VideoRecorder` wrapper (https://github.com/vwxyzjn/gym/blob/master/gym/wrappers/monitoring/video_recorder.py)?

**Comment by jkterry1** [comment]
Hey, a few things:

-This came up in past discussions, but just for my own knowledged why is dropping Python 3.6 support an issue? It was officially end of lifed 4 months ago, and numpy stopped supporting it like a year before that.
-If there's anything we can do to help regarding this upgrade, please let us know, we'd be happy to offer our thoughts or hop on a short call with you guys.

CC @pseudo-rnd-thoughts @arjun-kg

**Comment by gjoliver** [comment]
> Hey, a few things:
> 
> -This came up in past discussions, but just for my own knowledged why is dropping Python 3.6 support an issue? It was officially end of lifed 4 months ago, and numpy stopped supporting it like a year before that. -If there's anything we can do to help regarding this upgrade, please let us know, we'd be happy to offer our thoughts or hop on a short call with you guys.
> 
> CC @pseudo-rnd-thoughts @arjun-kg

We still have partners that are on 3.6.

**Comment by gjoliver** [comment]
actually, @jkterry1, out of curiosity, what features do you need that made you guys drop 3.6 support?

**Comment by duburcqa** [comment]
Python 3.6 is instead still widely used in the industry, and it is the default in Ubuntu 18 with April 2023 EOL. Plus, I see no real features brought by Python 3.7 that legitimate dropping support of Python 3.6. I think numpy is not a good example because it is already mature and only adding features that are not critical for users, so that it is completely acceptable to use a version outdated by several years. It is not the case for packages as gym or ray

**Comment by jkterry1** [comment]
There wasn't a hard requirement absolutely motivating dropping Python 3.6 support. We dropped Python 3.6 because all the new versions of all the packages we depend upon but one have dropped support for Python 3.6 (or are unmaintained/deprecated and are things we've been working to replace with new packages that have already dropped 3.6 support). Also to use Gym you usually have to use PyTorch or TensorFlow, both of which have dropped Python 3.6 support. 

Outside of RLlib, I'm only aware of two contexts where dropping Python 3.6 caused an issue for any users of Gym, and both of them just upgraded to 3.7 with Conda or a similar tool on the same machine without any issues and continued as normal.

**Comment by duburcqa** [comment]
In general in the industry python depencies are pinned so that it cannot break and only updated if there is a hard requirement. So I'm not surprise people are not complaining. They are just sticking to the old version that is working just fine for their needs. I know very well companies and laboratories in this situation. They are just not speaking up on GitHub.

File: rllib/agents/trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (avnishn) [reply rllib/agents/trainer.py:244]: Definitely not a bad idea -- I think itd be good to turn that into a good first issue and then have the community tackle it.

**Comment by sven1977** [comment]
We have fixed the (hopefully only) remaining issue with ray.remote vector environments. Waiting for tests to pass, then will merge. Again, fingers crossed :)
@jkterry1 @duburcqa 

Thanks @avnishn for taking the lead on this!

**Review by sven1977 — APPROVED** [review]
LGTM.
